### PR TITLE
Jonch/algoptminusglv

### PR DESCRIPTION
**Pull request by jon-chuang**

### CODE REVIEW

File: algebra-benches/src/macros/batch_arith.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+macro_rules! batch_arith {
```
Comment (kobigurk) [inline algebra-benches/src/macros/batch_arith.rs:1]: Doesn't this run only on nightly?
If yes, let's move these to a criterion benchmark.

File: algebra-core/src/biginteger/macros.rs [hunk 2]
```diff
@@ -198,6 +198,48 @@ macro_rules! bigint_impl {
...
+                    for j in 0..($num_limbs - i) {
+                        r[j + i] =
+                            arithmetic::mac_with_carry(r[j + i], this[i], other[j], &mut carry);
+                    }
+                }
+                Self::new(r)
+            }
+
+            #[inline]
+            fn from_slice(slice: &[u64]) -> Self {
```
Comment (kobigurk) [inline algebra-core/src/biginteger/macros.rs:236]: Do we want to return an error when the slice is too large?

File: algebra-core/src/biginteger/mod.rs [hunk 3]
```diff
@@ -107,8 +111,8 @@ pub trait BigInteger:
     /// Returns true iff this number is zero.
     fn is_zero(&self) -> bool;
 
-    /// Compute the number of bits needed to encode this number. Always a
-    /// multiple of 64.
+    /// Compute the exact number of bits needed to encode this number. Does not need
+    /// to be multiple of 64
     fn num_bits(&self) -> u32;
```
Comment (kobigurk) [inline algebra-core/src/biginteger/mod.rs:116]: Was it always like this? If no, people may be relying on it, so might be better to have a new method.

File: algebra-core/src/biginteger/mod.rs [hunk 4]
```diff
@@ -137,6 +141,18 @@ pub trait BigInteger:
         *self = Self::read(reader)?;
         Ok(())
     }
+
+    /// Takes two slices of u64 representing big integers and returns a bigger BigInteger
+    /// of type Self representing their product. Preferably used only for even NUM_LIMBS.
+    /// We require the invariant that this.len() == other.len() <= NUM_LIMBS / 2
+    fn mul_no_reduce(this: &[u64], other: &[u64]) -> Self;
```
Comment (kobigurk) [inline algebra-core/src/biginteger/mod.rs:148]: Amazing comments :)

File: algebra-core/src/biginteger/mod.rs [hunk 5]
```diff
@@ -137,6 +141,18 @@ pub trait BigInteger:
...
+
+    /// Similar to `mul_no_reduce` but accepts slices of with len == NUM_LIMBS
```
Comment (kobigurk) [inline algebra-core/src/biginteger/mod.rs]:
```suggestion
    /// Similar to `mul_no_reduce` but accepts slices of len == NUM_LIMBS
```

File: algebra-core/src/biginteger/mod.rs [hunk 6]
```diff
@@ -242,4 +258,7 @@ pub mod arithmetic {
 
         *carry = (tmp >> 64) as u64;
     }
+
+    // #[inline]
```
Comment (kobigurk) [inline algebra-core/src/biginteger/mod.rs]: Delete?

File: algebra-core/src/curves/batch_arith.rs [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+                .collect();
+
+            Self::batch_add_in_place(&mut bases, &mut add_ops[..], &index_add[..]);
+        }
+    }
+
+    /// Chunks vectorised instructions into a size that does not require
+    /// storing a lot of intermediate state
+
+    // Maybe put this as a helper function instead of in the trait?
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_arith.rs]: Let's put at as an issue rather than a comment?

File: algebra-core/src/curves/batch_arith.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+            } else {
+                chunk.to_vec()
+            };
+            res.push(chunk);
+        }
+        res
+    }
+}
+
+/// We make the syntax cleaner by defining corresponding trait and impl for [G]
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_arith.rs]: Change the comment to describe what it does better?

File: algebra-core/src/curves/batch_verify.rs [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+#[derive(Debug, Clone)]
+pub struct VerificationError;
+
+impl fmt::Display for VerificationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Verification Error. Not in subgroup")
+    }
+}
+
+// Only pass new_security_param if possibly recursing
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]: Maybe move the comment near `new_security_param`.

File: algebra-core/src/curves/batch_verify.rs [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    }
+    let mut buckets = batch_bucketed_add_split(num_buckets, points, &bucket_assign[..], 12);
+
+    // Check that all the buckets belong to the subgroup, either by calling
+    // the batch verify recusively, or by directly checking by multiplying by group order
+    // when the number of buckets is small enough
+    if num_buckets <= MAX_BUCKETS_FOR_FULL_CHECK || new_security_param == None {
+        // We use the batch scalar mul to check the subgroup condition if
+        // there are sufficient number of buckets
+        let verification_failure = if num_buckets >= 4096 {
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]: Is 4096 arbitrary or based on some benchmarks?

File: algebra-core/src/curves/batch_verify.rs [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+            cfg_chunks_mut!(buckets, 4096).for_each(|e| {
+                let length = e.len();
+                e[..].batch_scalar_mul_in_place::<<C::ScalarField as PrimeField>::BigInt>(
+                    &mut vec![<C::ScalarField as PrimeField>::Params::MODULUS.into(); length][..],
+                    1,
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]: Why 1?

File: algebra-core/src/curves/batch_verify.rs [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+            !buckets.iter().all(|&b| {
+                b.mul(<C::ScalarField as PrimeField>::Params::MODULUS) == C::Projective::zero()
+            })
+        };
+        if verification_failure {
+            return Err(VerificationError);
+        }
+    } else {
+        // Since !new_security_param.is_none():
+        let new_security_param = new_security_param.unwrap();
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]: Seems like this whole branch of code is only for recursion. Let's just remove it and create an issue.

File: algebra-core/src/curves/batch_verify.rs [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    let (num_buckets, num_rounds, _) = get_max_bucket(
+        security_param,
+        points.len(),
+        <C::ScalarField as PrimeField>::Params::MODULUS_BITS as usize,
+    );
+    run_rounds(points, num_buckets, num_rounds, None, rng)?;
+    Ok(())
+}
+
+/// Temporarily commented out until a fix can be found for the recursive version of the test
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]: Same comments as above about removing.

File: algebra-core/src/curves/batch_verify.rs [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+// ) -> Result<(), VerificationError> {
+//     // we add security for maximum depth, as recursive depth adds additional error to error bound
+//     let security_param = security_param + (log2(log2(security_param) as usize) as usize) + 1;
+//     let (num_buckets, num_rounds, new_security_param) =
+//         get_max_bucket(security_param, points.len(), 2);
+//     run_rounds(points, num_buckets, num_rounds, Some(new_security_param), rng)?;
+//     Ok(())
+// }
+
+pub fn batch_verify_in_subgroup_proj<C: ProjectiveCurve, R: Rng>(
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]: Same comments as above about removing (or moving to a different branch).

File: algebra-core/src/curves/batch_verify.rs [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+                return Err(VerificationError);
+            }
+        } else {
+            batch_verify_in_subgroup_proj(&buckets[..], new_security_param, rng)?;
+        }
+    }
+    Ok(())
+}
+
+// We get the greatest power of 2 number of buckets
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]: Rust comments

File: algebra-core/src/curves/bucketed_add.rs [hunk 16]
```diff
@@ -0,0 +1,256 @@
+use crate::{cfg_iter_mut, curves::BatchGroupArithmeticSlice, log2, AffineCurve, Vec};
+
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+#[cfg(feature = "std")]
+use std::collections::HashMap;
+
+const RATIO_MULTIPLIER: usize = 2;
+const BATCH_ADD_SIZE: usize = 4096;
```
Comment (kobigurk) [inline algebra-core/src/curves/bucketed_add.rs]: There was another place that references 4096, could they refer to the same constant?

File: algebra-core/src/curves/bucketed_add.rs [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+            elems[..].batch_add_in_place_same_slice(&instr_chunk[..]);
+        }
+    }
+
+    let zero = C::zero();
+    let mut res = vec![zero; buckets];
+
+    for bucket in 0..buckets {
+        if index[offset * bucket] > 1 {
+            panic!("Did not successfully reduce to_add");
```
Comment (kobigurk) [inline algebra-core/src/curves/bucketed_add.rs]: Kinda scary, can this even happen?

File: algebra-core/src/curves/bucketed_add.rs [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+        } else if index[offset * bucket] == 1 {
+            res[bucket] = elems[index[offset * bucket + 1] as usize];
+        }
+    }
+    res
+}
+
+#[cfg(not(feature = "std"))]
+pub fn batch_bucketed_add<C: AffineCurve>(
```
Comment (kobigurk) [inline algebra-core/src/curves/bucketed_add.rs:202]: I'm assuming you've looked at deduping all the common code and found it too cumbersome?

File: algebra/src/tests/curves.rs [hunk 19]
```diff
@@ -222,6 +237,319 @@ fn random_transformation_test<G: ProjectiveCurve>() {
...
+
+        let res1: Vec<C::Projective> = res1.iter().map(|&p| p.into()).collect();
+
+        for (i, (p1, p2)) in res1.iter().zip(res2).enumerate() {
+            assert_eq!(*p1, p2);
+        }
+    }
+}
+
+macro_rules! batch_verify_test {
```
Comment (kobigurk) [inline algebra/src/tests/curves.rs:448]: Are the commented out tests take too long to run?
Btw, the printing ones can remain, we usually don't see the prints.

File: algebra/src/tests/helpers.rs [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+use algebra_core::{
+    AffineCurve, BatchGroupArithmeticSlice, BigInteger64, ProjectiveCurve, UniformRand,
+};
+use rand::{distributions::Uniform, prelude::Distribution};
+use rand_xorshift::XorShiftRng;
+
+#[cfg(feature = "parallel_random_gen")]
+use rayon::prelude::*;
+
+pub fn create_pseudo_uniform_random_elems<C: AffineCurve>(
```
Comment (kobigurk) [inline algebra/src/tests/helpers.rs]: Nit: this can still receive Rng.

File: algebra/sudo [hunk 21]
```diff
@@ -0,0 +1 @@
+off tee /sys/devices/system/cpu/smt/control
```
Comment (kobigurk) [inline algebra/sudo]: Hello there file :)

File: scripts/glv_lattice_basis/src/arithmetic.rs [hunk 22]
```diff
@@ -0,0 +1,34 @@
+use algebra_core::biginteger::BigInteger;
```
Comment (kobigurk) [inline scripts/glv_lattice_basis/src/arithmetic.rs]: Seems like GLV stuff did sneak in :(
Can you remove?

Also all the GLV bools.

File: algebra-core/src/curves/glv.rs [hunk 23]
```diff
@@ -0,0 +1,85 @@
+use crate::{biginteger::BigInteger, ModelParameters, PrimeField};
```
Comment (kobigurk) [inline algebra-core/src/curves/glv.rs]: See other GLV comment

File: algebra-core/src/curves/models/twisted_edwards_extended.rs [hunk 24]
```diff
@@ -165,6 +165,129 @@ impl<P: Parameters> AffineCurve for GroupAffine<P> {
...
+            &mut bases.to_vec()[..],
+            &index
+                .iter()
+                .map(|&x| (x, x))
+                .collect::<Vec<(usize, usize)>>()[..],
+        );
+    }
+
+    // Total cost: 12 mul. Projective formulas: 11 mul.
+    fn batch_add_in_place_same_slice(bases: &mut [Self], index: &[(usize, usize)]) {
```
Comment (kobigurk) [inline algebra-core/src/curves/models/twisted_edwards_extended.rs]: Seems like a lot of shared code between this method and the next. Can we dedup somehow?

File: algebra-core/src/msm/variable_base.rs [hunk 25]
```diff
@@ -39,7 +41,8 @@ impl VariableBaseMSM {
             .map(|w_start| {
                 let mut res = zero;
                 // We don't need the "zero" bucket, so we only have 2^c - 1 buckets
-                let mut buckets = vec![zero; (1 << c) - 1];
+                let log2_n_bucket = if (w_start % c) != 0 { w_start % c } else { c };
```
Comment (kobigurk) [inline algebra-core/src/msm/variable_base.rs:44]: Interesting, why this change?

File: algebra/src/tests/curves.rs [hunk 26]
```diff
@@ -222,6 +237,319 @@ fn random_transformation_test<G: ProjectiveCurve>() {
     }
 }
 
+pub fn random_batch_doubling_test<G: ProjectiveCurve>() {
```
Comment (kobigurk) [inline algebra/src/tests/curves.rs:240]: Great tests all around!

**Review by kobigurk — CHANGES_REQUESTED** [review]

**Review by kobigurk — CHANGES_REQUESTED** [review]
This is awesome over all!

File: algebra-core/src/biginteger/mod.rs (hunk 3 above)
Comment (jon-chuang) [reply algebra-core/src/biginteger/mod.rs:116]: It was always like this, the comments were wrong

File: algebra-core/src/biginteger/macros.rs (hunk 2 above)
Comment (jon-chuang) [reply algebra-core/src/biginteger/macros.rs:236]: Might actually help the compiler by not doing bounds check for every slice access

File: algebra-core/src/curves/bucketed_add.rs (hunk 17 above)
Comment (jon-chuang) [reply algebra-core/src/curves/bucketed_add.rs]: Technically, if the code is correct, no. It was used for debugging. And if the code is wrong in some subtle way, at least if could catch a silent error. I'll change to a debug assert so the compiler optimizes it away

File: algebra-core/src/curves/bucketed_add.rs (hunk 18 above)
Comment (jon-chuang) [reply algebra-core/src/curves/bucketed_add.rs:142]: I noticed that one can indeed dedup. I used macros as defining a private trait was to cumbersome

File: algebra-core/src/curves/models/twisted_edwards_extended.rs (hunk 24 above)
Comment (jon-chuang) [reply algebra-core/src/curves/models/twisted_edwards_extended.rs]: I noticed that one can indeed dedup. I used macros as defining a private fn in a trait was too cumbersome

File: algebra-core/src/msm/variable_base.rs (hunk 25 above)
Comment (jon-chuang) [reply algebra-core/src/msm/variable_base.rs:44]: it probably doesn't make a big difference. But still, let's say that you had just a few bits for the last window decomposition for the MSM. It could help a little with the memory allocation. 

Its just slightly more correct, anyway, if a nitpick

File: algebra-core/src/curves/bucketed_add.rs (hunk 18 above)
Comment (jon-chuang) [reply algebra-core/src/curves/bucketed_add.rs:202]: there's not much to dedup.  Most of the code is in creating the index for the in place addition tree. And they are completely different approaches.

File: algebra/src/tests/curves.rs (hunk 19 above)
Comment (jon-chuang) [reply algebra/src/tests/curves.rs:448]: The test below is just non deterministic and only works for small security parameters. It's an anti-test. I'd like to leave it there in case anyone wants to do sanity checking.

The recursive stuff I will put in an issue and we can add it back.

File: algebra-core/src/curves/bucketed_add.rs (hunk 16 above)
Comment (jon-chuang) [reply algebra-core/src/curves/bucketed_add.rs]: I've made it a common constant in batch arith.rs and add some comments

File: algebra-core/src/biginteger/macros.rs (hunk 2 above)
Comment (jon-chuang) [reply algebra-core/src/biginteger/macros.rs:236]: Oh I was referring to mul_no_reduce for bounds checking. For this one, I guess it makes no difference. It ought to be clear how many limbs a biginteger type has so it's easy to diagnose when it's suitable or not.

**Comment by jon-chuang** [comment]
sorry, I did a hard fork of the wrong parent branch. So it wiped out the ability to compare changes since the review comments. The only comments I don't think have been addressed yet are the criterion based tests. It's a lot of work to overhaul the whole thing so I suggest it be done in a future PR.

Further, I don't see the issue with benching on nightly, and recall criterion was unsuitable for ops on the order of < 1us due to the long excessively long benching times, without tweaking.

File: algebra-core/src/curves/models/twisted_edwards_extended.rs (hunk 24 above)
Comment (kobigurk) [reply algebra-core/src/curves/models/twisted_edwards_extended.rs]: I think macros is a good way, do you want to do that change?

File: algebra-core/src/msm/variable_base.rs (hunk 25 above)
Comment (kobigurk) [reply algebra-core/src/msm/variable_base.rs:44]: Oh, I see, that makes sense.

**Comment by kobigurk** [comment]
> sorry, I did a hard fork of the wrong parent branch. So it wiped out the ability to compare changes since the review comments. The only comments I don't think have been addressed yet are the criterion based tests. It's a lot of work to overhaul the whole thing so I suggest it be done in a future PR.
> 
> Further, I don't see the issue with benching on nightly, and recall criterion was unsuitable for ops on the order of < 1us due to the long excessively long benching times, without tweaking.

Yeah, we can definitely leave that for a future PR. 
Benching with nightly slightly makes me nervous since that's not what we'll be running eventually.

File: algebra-core/src/curves/bucketed_add.rs (hunk 18 above)
Comment (kobigurk) [reply algebra-core/src/curves/bucketed_add.rs:202]: I guess you're right, the common code is not that large.

File: algebra/src/tests/curves.rs (hunk 19 above)
Comment (kobigurk) [reply algebra/src/tests/curves.rs:448]: Got it. Maybe we can put a comment about that?

File: algebra-core/src/biginteger/macros.rs [hunk 27]
```diff
@@ -218,8 +218,8 @@ macro_rules! bigint_impl {
 
             #[inline]
             fn mul_no_reduce_lo(this: &[u64], other: &[u64]) -> Self {
-                debug_assert!(this.len() == $num_limbs);
-                debug_assert!(this.len() == other.len());
+                assert!(this.len() == $num_limbs);
+                assert!(other.len() == $num_limbs);
```
Comment (kobigurk) [inline algebra-core/src/biginteger/macros.rs:222]: Can someone crash someone else's system using this assert?

File: algebra/src/tests/curves.rs (hunk 19 above)
Comment (kobigurk) [reply algebra/src/tests/curves.rs:448]: I've just seen you did, thanks!

File: algebra/src/ed_on_cp6_782/curves/tests.rs [hunk 28]
```diff
@@ -12,6 +12,7 @@ use crate::tests::{curves::*, groups::*};
 fn test_projective_curve() {
     curve_tests::<EdwardsProjective>();
 
+    println!("EDwards tests");
```
Comment (kobigurk) [inline algebra/src/ed_on_cp6_782/curves/tests.rs]:
```suggestion
    println!("Edwards tests");
```

File: algebra-core/src/curves/models/twisted_edwards_extended.rs (hunk 24 above)
Comment (kobigurk) [reply algebra-core/src/curves/models/twisted_edwards_extended.rs]: I've just seen you did, thanks!

File: algebra-benches/src/macros/batch_arith.rs (hunk 1 above)
Comment (kobigurk) [reply algebra-benches/src/macros/batch_arith.rs:1]: Agreed to do on another PR.

File: algebra-core/src/curves/batch_verify.rs [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+impl fmt::Display for VerificationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Verification Error. Not in subgroup")
+    }
+}
+
+fn verify_points<C: AffineCurve, R: Rng>(
+    points: &[C],
+    num_buckets: usize,
+    _new_security_param: Option<usize>, // Only pass new_security_param if possibly recursing (future PRs)
```
Comment (kobigurk) [inline algebra-core/src/curves/batch_verify.rs]:
```suggestion
    _new_security_param: Option<usize>, // Only pass new_security_param if possibly recursing
```

File: algebra-core/src/biginteger/macros.rs (hunk 2 above)
Comment (Pratyush) [reply algebra-core/src/biginteger/macros.rs:236]: I've discovered that when one can return an `Option`, it's probably the right thing to do (eg: with `from_repr`)

File: algebra-core/src/curves/batch_arith.rs [hunk 30]
```diff
@@ -0,0 +1,294 @@
...
+        index_code as usize >> ENDO_CODING_BITS,
+        index_code as u8 % 4,
+    )
+}
+
+pub trait BatchGroupArithmetic
+where
+    Self: Sized + Clone + Copy + Zero + Neg<Output = Self>,
+{
+    type BBaseField: Field;
```
Comment (Pratyush) [inline algebra-core/src/curves/batch_arith.rs:25]: Is this meant to be `BaseField`?

File: algebra-core/src/curves/batch_arith.rs [hunk 31]
```diff
@@ -0,0 +1,294 @@
...
+        let window_size: i16 = 1 << (w + 1);
+        let half_window_size: i16 = 1 << w;
+
+        let mut op_code_vectorised =
+            Vec::<Vec<Option<i16>>>::with_capacity(scalars[0].as_ref().len() * 64);
+
+        let mut all_none = false;
+
+        match negate {
+            None => {
```
Comment (Pratyush) [inline algebra-core/src/curves/batch_arith.rs]: Probably these two match branches can be unified, perhaps with something like the following:

```rust
negate.unwrap_or((), |b| assert_eq!(b.len(), scalars.len()); // precompute bounds check
let bools = (0..scalars.len()).map(|i| negate.map_or(false, |b| b[i]]);
```

When `negate` is `None`, I expect this to optimize out unused branches without issues.

**Comment by kobigurk** [comment]
Since this has essentially become "stale" and the GLV branch is the main one, should we close this?
